apps/prairielearn/src/lib/groups.sql

```diff
@@ -193,65 +188,50 @@ WITH
     WHERE
       user_id = $user_id
       AND group_id = $group_id
-  ),
-  deleted_group_user_roles AS (
-    DELETE FROM group_user_roles
-    WHERE
-      user_id = $user_id
-      AND group_id = $group_id
   )
 INSERT INTO
   group_logs (authn_user_id, user_id, group_id, action)
 VALUES
   ($authn_user_id, $user_id, $group_id, 'leave');
 
--- BLOCK reassign_group_roles_after_leave
+-- BLOCK update_group_roles
 WITH
-  deleted_group_users_roles AS (
-    DELETE FROM group_user_roles
-    WHERE
-      group_id = $group_id
-  ),
```

Comment:
The strategy previously used here was to delete all group user roles, then insert them back with the new settings. The problem here was that the CTEs did not necessarily run in the order listed, so deletion could happen after insertion, causing a unique key conflict. I have changed the query to only delete user roles that are not in the input array.